044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,248 @@
+# Cluster Wide Volume Usage Quota Management
+
+Extend the static quota mechanism to throttle message production in a broker if any active broker in the cluster is running out of disk.
+
+Deprecate the current implementation that considers only aggregate local volume usage.
+
+- [Current situation](#current-situation)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+  - [High Level Changes](#high-level-changes)
+    - [Volume](#volume)
+    - [Volume Source](#volume-source)
+    - [Quota Source](#quota-source)
+    - [Throttle Factor](#throttle-factor)
+    - [Cluster Volume Source](#cluster-volume-source)
+      - [Admin Client Configuration](#admin-client-configuration)
+      - [Limit Type Configuration](#limit-type-configuration)
+    - [Fallback Throttle Factor](#fallback-throttle-factor)
+- [Metrics](#metrics)
+- [Rejected Alternatives](#rejected-alternatives)
+- [Affected Projects](#affected-projects)
+- [Compatibility](#compatibility)
+
+## Current situation
+
+The [kafka-static-quota-plugin](https://github.com/strimzi/kafka-quotas-plugin) applies byte-rate limits on client connections to
+individual brokers (thresholds derived from storage limits), slowing producers down when consumed storage is over the soft limit (but below the hard limit) and
+effectively pausing publication when reaching or exceeding the hard limit. This largely prevents out of disk scenarios when topics
+are replicated to all brokers thus there is largely similar disk usage between all brokers. Assuming all the brokers
+have similar disk usage levels, they will all apply rate limits at similar times and levels, effectively giving cluster-wide out-of-storage protection.
+
+This however provides limited protection to clusters with un-even distribution of topics and thus storage usage.
+Additionally, it ties the plug-in directly to storage usage, there are other factors which users may wish to respond to.
+
+As clusters scale up the likelihood of even topic distribution drops. When topics are not evenly distributed it
+is possible for replication from broker `1` to broker `2` will cause broker `2` to consume all available storage
+without triggering throttling of clients, as broker `1` disk usage remains acceptable.
+
+Addressing the effects of uneven topic distribution sounds like it should come under [KIP-73](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas). Unfortunately replication quotas
+are designed to manage the additional network load of migrating replicas between brokers, which does not address client publication
+leading to out of disk conditions through replication.
```

Comment:
`Unfortunately replication quotas are designed to manage the additional network load of migrating replicas between brokers, which does not address client publication` <-- Why does it not address client publication issue here? Could you elaborate more? IMO, it works on throttling follower fetching from leaders.

Reply:
I was trying to avoid getting into the details in the Proposal but can elaborate further.

The problem with KIP-73 is it targets specific topic partitions. Which makes a lot of sense when re-distributing topic partitions around a cluster, however from the quota plug-in we have no way to determine which topic partitions are affected and would thus need throttling. 

On a slightly more fundamental level if we don't reduce the flow at the front door (i.e. throttle producers) we can't reduce the demand on storage. If we just throttled replication to specific brokers we end up with all sorts of nasty latency surprises in brokers where messages will be stalled replicating to a specific broker and thus delay acks to back to clients.